### PR TITLE
feat: Initial support for visualization on Kibana 7.10

### PR DESCRIPTION
**Pull request by dupuyjs**
# feat: Initial support for visualization on Kibana 7.10

Implement changes to support base requirements for date histogram visualization.

### CODE REVIEW

File: K2Bridge/JsonConverters/LeafAggregationConverter.cs [hunk 1]
```diff
@@ -25,6 +25,9 @@ internal class LeafAggregationConverter : ReadOnlyJsonConverter
             var jo = JObject.Load(reader);
             var first = (JProperty)jo.First;
```
Comment (dupuyjs) [inline K2Bridge/JsonConverters/LeafAggregationConverter.cs:26]: We should probably not assume that PrimaryAggregation will be always on the first node. If "aggs" is the first node, its still a valid json and will break the code.

File: K2Bridge/Models/Response/DateHistogramBucket.cs [hunk 2]
```diff
@@ -30,5 +33,11 @@ public class DateHistogramBucket : IBucket
         /// </summary>
         [JsonProperty("key_as_string")]
         public string KeyAsString { get; set; }
+
+        /// <summary>
+        /// Gets or Sets fdsfd.
+        /// </summary>
+        [JsonIgnore]
+        public Dictionary<string, List<long>> Aggs { get; set; }
```
Comment (tomconte) [inline K2Bridge/Models/Response/DateHistogramBucket.cs:41]: This should be a floating-point type. ES does return floats for aggregations like average.

File: K2Bridge/Factories/BucketFactory.cs [hunk 3]
```diff
@@ -23,16 +23,32 @@ public static DateHistogramBucket CreateFromDataRow(DataRow row)
...
+            var clmns = row.Table.Columns;
+            foreach (DataColumn clmn in clmns)
+            {
+                // todo: fix
+                if (clmn.ColumnName == "count_" || clmns.IndexOf(clmn) == (int)DateHistogramBucketColumnNames.Timestamp)
+                {
+                    continue;
+                }
+
+                dhb.Aggs[clmn.ColumnName.Substring(1)] = new System.Collections.Generic.List<long>() { Convert.ToInt64(row[clmn.ColumnName]) };
```
Comment (tomconte) [inline K2Bridge/Factories/BucketFactory.cs:48]: This should convert to floating-point and not integer.

File: K2Bridge/Factories/BucketFactory.cs [hunk 4]
```diff
@@ -23,16 +23,32 @@ public static DateHistogramBucket CreateFromDataRow(DataRow row)
...
+            var count = row["count_"];
             var dateBucket = (DateTime)timestamp;
 
-            return new DateHistogramBucket
+            var dhb = new DateHistogramBucket
             {
                 DocCount = Convert.ToInt32(count),
                 Key = TimeUtils.ToEpochMilliseconds(dateBucket),
                 KeyAsString = dateBucket.ToString("yyyy-MM-ddTHH:mm:ss.fffK"),
+                Aggs = new System.Collections.Generic.Dictionary<string, System.Collections.Generic.List<long>>(),
```
Comment (dupuyjs) [inline K2Bridge/Factories/BucketFactory.cs]: I suppose long is not appropriate for some aggregations, double seems more appropriate.

File: K2Bridge/Factories/BucketFactory.cs (hunk 3 above)
Comment (dupuyjs) [inline K2Bridge/Factories/BucketFactory.cs]: I suppose long is not appropriate for some aggregations, double seems more appropriate.

File: K2Bridge/JsonConverters/LeafAggregationConverter.cs (hunk 1 above)
Comment (tomconte) [reply K2Bridge/JsonConverters/LeafAggregationConverter.cs:26]: I have finally confirmed that Kibana does send the primary aggregation as the first node; I had incorrectly used Jthe SON reformatted by MITMproxy, that had modified the order of elements in the JSON...

That said, the comment is still valid, we can't say for sure that the primary aggregation will always be the first element.

File: K2Bridge/JsonConverters/AggregationConverter.cs [hunk 5]
```diff
@@ -24,16 +24,24 @@ internal class AggregationConverter : ReadOnlyJsonConverter
...
+            var parent = path[path.Length - 1];
+
             var obj = new Aggregation
             {
                 PrimaryAggregation = jo.ToObject<LeafAggregation>(serializer),
             };
 
-            var aggsObject = jo["aggs"];
-            if (aggsObject != null)
+            if (obj.PrimaryAggregation != null)
```
Comment (dupuyjs) [inline K2Bridge/JsonConverters/AggregationConverter.cs:35]: I could also probably check that parent is not null or empty.

File: K2Bridge/Factories/BucketFactory.cs [hunk 6]
```diff
@@ -23,16 +23,32 @@ public static DateHistogramBucket CreateFromDataRow(DataRow row)
         {
             Ensure.IsNotNull(row, nameof(row));
 
+            // fix column names
             var timestamp = row[(int)DateHistogramBucketColumnNames.Timestamp];
-            var count = row[(int)DateHistogramBucketColumnNames.Count];
+            var count = row["count_"];
```
Comment (dupuyjs) [inline K2Bridge/Factories/BucketFactory.cs:28]: Do we need to add "count_" in a constant string definition or in DateHistogramsBucketColumnNames ?

File: K2Bridge/Models/Response/Aggregations.cs [hunk 7]
```diff
@@ -4,17 +4,23 @@
...
     public class Aggregations
     {
         /// <summary>
         /// Gets or sets aggregation value.
         /// </summary>
-        [JsonProperty("2")]
         public BucketsCollection Collection { get; set; } = new BucketsCollection();
+
+        /// <summary>
+        /// Gets or Sets sa.
```
Comment (dupuyjs) [inline K2Bridge/Models/Response/Aggregations.cs:22]: Need to update the property summary.

File: K2Bridge/Models/Response/DateHistogramBucket.cs [hunk 8]
```diff
@@ -30,5 +33,11 @@ public class DateHistogramBucket : IBucket
         /// </summary>
         [JsonProperty("key_as_string")]
         public string KeyAsString { get; set; }
+
+        /// <summary>
+        /// Gets or Sets fdsfd.
```
Comment (dupuyjs) [inline K2Bridge/Models/Response/DateHistogramBucket.cs:38]: Need to update the property summary.

File: K2Bridge/JsonConverters/ResponseAggregationsConverter.cs [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+    using System.Collections.Generic;
+    using K2Bridge.Models.Request.Aggregations;
+    using K2Bridge.Models.Response;
+    using Newtonsoft.Json;
+    using Newtonsoft.Json.Linq;
+
+    /// <summary>
+    /// A converter able to deserialize the aggs element in an Elasticsearch query to <see cref="Aggregation"/>.
+    /// </summary>
+    internal class ResponseAggregationsConverter : JsonConverter
```
Comment (dupuyjs) [inline K2Bridge/JsonConverters/ResponseAggregationsConverter.cs:17]: To be aligned with other converters. Should we create a WriteOnlyJsonConverter class and inherit from it?

File: K2Bridge/JsonConverters/SubAggsConverter.cs [hunk 10]
```diff
@@ -0,0 +1,71 @@
...
+        }
+
+        /// <inheritdoc/>
+        public override void WriteJson(JsonWriter writer, object value, JsonSerializer serializer)
+        {
+            var buck = value as DateHistogramBucket;
+            var aggs = buck.Aggs;
+
+            // Dictionary<string, List<string>>;
+            // writer.WriteStartArray();
```
Comment (dupuyjs) [inline K2Bridge/JsonConverters/SubAggsConverter.cs:38]: Need to remove these comments.

File: K2Bridge/Factories/BucketFactory.cs (hunk 6 above)
Comment (tomconte) [reply K2Bridge/Factories/BucketFactory.cs:28]: I would do it for consistency.

File: K2Bridge/JsonConverters/ResponseAggregationsConverter.cs (hunk 9 above)
Comment (tomconte) [reply K2Bridge/JsonConverters/ResponseAggregationsConverter.cs:17]: I guess it would be nice, again for consistency with `ReadOnlyJsonConverter`...

**Review by tomconte — CHANGES_REQUESTED** [review]
Looks good to me! Just suggested a couple of comment updates.

File: K2Bridge/JsonConverters/AggregationConverter.cs (hunk 5 above)
Comment (dupuyjs) [reply K2Bridge/JsonConverters/AggregationConverter.cs]: Fixed.

File: K2Bridge/Factories/BucketFactory.cs (hunk 6 above)
Comment (dupuyjs) [reply K2Bridge/Factories/BucketFactory.cs]: Added count_ to DateHistogramsBucketColumnNames. As DateHistogramsBucketColumnNames contains now int and string, updated the enumeration to a class with readonly static property.

File: K2Bridge/Models/Response/Aggregations.cs (hunk 7 above)
Comment (dupuyjs) [reply K2Bridge/Models/Response/Aggregations.cs]: Fixed.

File: K2Bridge/JsonConverters/ResponseAggregationsConverter.cs (hunk 9 above)
Comment (dupuyjs) [reply K2Bridge/JsonConverters/ResponseAggregationsConverter.cs]: Fixed. And Moved ReadOnlyJsonConverter and WriteOnlyJsonConverter to a base folder.

File: K2Bridge/JsonConverters/SubAggsConverter.cs (hunk 10 above)
Comment (dupuyjs) [reply K2Bridge/JsonConverters/SubAggsConverter.cs]: Fixed.

File: K2Bridge/Models/Response/DateHistogramBucket.cs (hunk 8 above)
Comment (dupuyjs) [reply K2Bridge/Models/Response/DateHistogramBucket.cs]: Fixed.

File: K2Bridge/Models/Request/Aggregations/MaxAggregation.cs [hunk 11]
```diff
@@ -0,0 +1,23 @@
...
+// See LICENSE file in the project root for full license information.
+
+namespace K2Bridge.Models.Request.Aggregations
+{
+    using K2Bridge.JsonConverters;
+    using K2Bridge.Visitors;
+    using Newtonsoft.Json;
+
+    /// <summary>
+    /// A single-value metrics aggregation that computes the average of numeric values that are extracted from the aggregated documents.
```
[the remark this answers is not in the document]
Comment (dupuyjs) [reply K2Bridge/Models/Request/Aggregations/MaxAggregation.cs]: Fixed.

File: K2Bridge/Visitors/IVisitor.cs [hunk 12]
```diff
@@ -79,6 +79,12 @@ internal interface IVisitor
         /// <param name="cardinalityAggregation">The Cardinality Aggregations to visit.</param>
         void Visit(CardinalityAggregation cardinalityAggregation);
 
+        /// <summary>
+        /// Accepts a given visitable object and builds a Kusto query.
+        /// </summary>
+        /// <param name="maxAggregation">The Cardinality Aggregations to visit.</param>
```
[the remark this answers is not in the document]
Comment (dupuyjs) [reply K2Bridge/Visitors/IVisitor.cs]: Fixed.

**Review by tomconte — APPROVED** [review]

File: K2Bridge.Tests.UnitTests/KustoDAL/ElasticResponseTests.cs [hunk 13]
```diff
@@ -103,7 +103,7 @@ public string MapRowsToHits_WhenMultiWordDifferentCapsCapialLettersInQueryString
         }
 
         [TestCase(ExpectedResult =
-            "{\"responses\":[{\"aggregations\":{\"2\":{\"buckets\":[]}},\"took\":0,\"timed_out\":false,\"_shards\":{\"total\":1,\"successful\":1,\"skipped\":0,\"failed\":0},\"hits\":{\"total\":{\"relation\":\"eq\",\"value\":0},\"max_score\":null,\"hits\":[]},\"status\":200}],\"took\":0}")]
+            "{\"responses\":[{\"aggregations\":{\"\":{\"buckets\":[]}},\"took\":0,\"timed_out\":false,\"_shards\":{\"total\":1,\"successful\":1,\"skipped\":0,\"failed\":0},\"hits\":{\"total\":{\"relation\":\"eq\",\"value\":0},\"max_score\":null,\"hits\":[]},\"status\":200}],\"took\":0}")]
```
Comment (eladiw) [inline K2Bridge.Tests.UnitTests/KustoDAL/ElasticResponseTests.cs:106]: an empty key looks wrong. I guess the input value should be changed and then we should expect to have a value in this key

File: K2Bridge.Tests.UnitTests/Models/Response/DateHistogramBucketTests.cs [hunk 14]
```diff
@@ -17,11 +17,11 @@ public void Create_WithDateTime_ReturnsISOString()
         {
             // Arrange
             DataTable table = new DataTable();
-            table.Columns.Add("Timestamp", typeof(DateTime)).DateTimeMode = DataSetDateTime.Utc;
-            table.Columns.Add("Count", typeof(int));
+            table.Columns.Add("timestamp", typeof(DateTime)).DateTimeMode = DataSetDateTime.Utc;
```
Comment (eladiw) [inline K2Bridge.Tests.UnitTests/Models/Response/DateHistogramBucketTests.cs:20]: why did we have to change these column names?

File: K2Bridge/Factories/BucketFactory.cs [hunk 15]
```diff
@@ -24,15 +24,30 @@ public static DateHistogramBucket CreateFromDataRow(DataRow row)
...
                 Key = TimeUtils.ToEpochMilliseconds(dateBucket),
                 KeyAsString = dateBucket.ToString("yyyy-MM-ddTHH:mm:ss.fffK"),
+                Aggs = new System.Collections.Generic.Dictionary<string, System.Collections.Generic.List<double>>(),
             };
+
+            var clmns = row.Table.Columns;
+            foreach (DataColumn clmn in clmns)
+            {
+                // todo: fix
+                if (clmn.ColumnName == DateHistogramBucketColumnNames.Count || clmns.IndexOf(clmn) == (int)DateHistogramBucketColumnNames.Timestamp)
```
Comment (eladiw) [inline K2Bridge/Factories/BucketFactory.cs:42]: probably just one of the if condition is needed

File: K2Bridge/JsonConverters/Base/WriteOnlyJsonConverter.cs [hunk 16]
```diff
@@ -0,0 +1,39 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT license.
+// See LICENSE file in the project root for full license information.
+
+namespace K2Bridge.JsonConverters.Base
```
Comment (eladiw) [inline K2Bridge/JsonConverters/Base/WriteOnlyJsonConverter.cs:5]: Is this file needed?

**Review by eladiw — CHANGES_REQUESTED** [review]
Some comments

File: K2Bridge.Tests.UnitTests/Models/Response/DateHistogramBucketTests.cs (hunk 14 above)
Comment (dupuyjs) [reply K2Bridge.Tests.UnitTests/Models/Response/DateHistogramBucketTests.cs:20]: It was not really needed, I updated these columns because names were in lowercase in other tests - and adx returns the columns in lowercase too. Anyway, tests were working fine before, and still works after.

File: K2Bridge/Factories/BucketFactory.cs [hunk 17]
```diff
@@ -24,15 +24,30 @@ public static DateHistogramBucket CreateFromDataRow(DataRow row)
...
                 DocCount = Convert.ToInt32(count),
                 Key = TimeUtils.ToEpochMilliseconds(dateBucket),
                 KeyAsString = dateBucket.ToString("yyyy-MM-ddTHH:mm:ss.fffK"),
+                Aggs = new System.Collections.Generic.Dictionary<string, System.Collections.Generic.List<double>>(),
             };
+
+            var clmns = row.Table.Columns;
+            foreach (DataColumn clmn in clmns)
+            {
+                // todo: fix
```
[the remark this answers is not in the document]
Comment (dupuyjs) [reply K2Bridge/Factories/BucketFactory.cs:41]: I don't think so, I will discuss with the point with @eladiw.

File: K2Bridge.Tests.UnitTests/KustoDAL/ElasticResponseTests.cs (hunk 13 above)
Comment (dupuyjs) [reply K2Bridge.Tests.UnitTests/KustoDAL/ElasticResponseTests.cs:106]: Added a bug to manage this point later.

File: K2Bridge/KustoDAL/KustoResponseParser.cs [hunk 18]
```diff
@@ -166,11 +166,13 @@ private void ReportQueryExecutionMetrics(KustoResponseDataSet kustoResponseDataS
             {
                 Logger.LogTrace("Parsing aggregations");
 
+                response.AddParentToAgg(parsedKustoResponse[AggregationTableName].TableData.Columns[0].ColumnName.Substring(1));
```
[the remark this answers is not in the document]
Comment (dupuyjs) [reply K2Bridge/KustoDAL/KustoResponseParser.cs:169]: Added additional comments in code.

File: K2Bridge/Factories/BucketFactory.cs (hunk 15 above)
Comment (dupuyjs) [reply K2Bridge/Factories/BucketFactory.cs:42]: Add a task to check if we can refactor this code.

File: K2Bridge/JsonConverters/AggregationConverter.cs [hunk 19]
```diff
@@ -23,17 +24,25 @@ internal class AggregationConverter : ReadOnlyJsonConverter
             JsonSerializer serializer)
         {
             var jo = JObject.Load(reader);
+            var aggs = jo["aggs"];
+
+            var path = reader.Path.Split(".");
+            var parent = path[path.Length - 1];
```
[the remark this answers is not in the document]
Comment (dupuyjs) [reply K2Bridge/JsonConverters/AggregationConverter.cs]: Fixed.

File: K2Bridge/Factories/BucketFactory.cs (hunk 17 above)
[the remark this answers is not in the document]
Comment (dupuyjs) [reply K2Bridge/Factories/BucketFactory.cs]: Comment has been removed.

File: K2Bridge.Tests.UnitTests/KustoDAL/ElasticResponseTests.cs [hunk 20]
```diff
@@ -102,8 +102,10 @@ public string MapRowsToHits_WhenMultiWordDifferentCapsCapialLettersInQueryString
                 "somevalue1 somevalue2");
         }
 
+        // TODO: An empty elastic response should probably not contains any aggregations.
```
Comment (eladiw) [inline K2Bridge.Tests.UnitTests/KustoDAL/ElasticResponseTests.cs:105]: can you add the issue link?

File: K2Bridge.Tests.UnitTests/KustoDAL/ElasticResponseTests.cs (hunk 13 above)
Comment (eladiw) [reply K2Bridge.Tests.UnitTests/KustoDAL/ElasticResponseTests.cs:108]: can you add the link?

File: K2Bridge/Factories/BucketFactory.cs [hunk 21]
```diff
@@ -23,16 +23,32 @@ public static DateHistogramBucket CreateFromDataRow(DataRow row)
         {
             Ensure.IsNotNull(row, nameof(row));
 
+            // TODO: timestamp is always the first row (probably named _2), and count will be named _count
```
Comment (eladiw) [inline K2Bridge/Factories/BucketFactory.cs:26]: can you add a link to the issue?

File: K2Bridge.Tests.UnitTests/KustoDAL/ElasticResponseTests.cs (hunk 20 above)
Comment (dupuyjs) [reply K2Bridge.Tests.UnitTests/KustoDAL/ElasticResponseTests.cs:105]: https://dev.azure.com/CSECodeHub/476026%20-%20K2Bridge%20OSS%20-%20Visualization%20Capabilities/_workitems/edit/15049

File: K2Bridge/Factories/BucketFactory.cs (hunk 21 above)
Comment (dupuyjs) [reply K2Bridge/Factories/BucketFactory.cs:26]: https://dev.azure.com/CSECodeHub/476026%20-%20K2Bridge%20OSS%20-%20Visualization%20Capabilities/_workitems/edit/15050

Comment (dupuyjs) [reply K2Bridge/Factories/BucketFactory.cs:26]: Link to the issue added to code.

File: K2Bridge.Tests.UnitTests/KustoDAL/ElasticResponseTests.cs (hunk 20 above)
Comment (dupuyjs) [reply K2Bridge.Tests.UnitTests/KustoDAL/ElasticResponseTests.cs:105]: Link to the issue added to code.

File: K2Bridge.Tests.UnitTests/KustoDAL/ElasticResponseTests.cs (hunk 13 above)
Comment (dupuyjs) [reply K2Bridge.Tests.UnitTests/KustoDAL/ElasticResponseTests.cs:109]: Link to the issue added to code.

File: K2Bridge.Tests.UnitTests/KustoDAL/ElasticResponseTests.cs [hunk 22]
```diff
@@ -102,8 +102,11 @@ public string MapRowsToHits_WhenMultiWordDifferentCapsCapialLettersInQueryString
...
+        // And here there is no key for the aggregations dictionary. This is not a valid json.
+        // https://dev.azure.com/CSECodeHub/476026%20-%20K2Bridge%20OSS%20-%20Visualization%20Capabilities/_workitems/edit/15049
```
Comment (eladiw) [inline K2Bridge.Tests.UnitTests/KustoDAL/ElasticResponseTests.cs:107]:
```suggestion
        // workitem: 15049
```

File: K2Bridge/Factories/BucketFactory.cs [hunk 23]
```diff
@@ -23,16 +23,33 @@ public static DateHistogramBucket CreateFromDataRow(DataRow row)
...
+            // we currently mix index and column names, need to check if we can enhance this logic
+            // https://dev.azure.com/CSECodeHub/476026%20-%20K2Bridge%20OSS%20-%20Visualization%20Capabilities/_workitems/edit/15050
```
Comment (eladiw) [inline K2Bridge/Factories/BucketFactory.cs:28]:
```suggestion
            // workitem 15050
```

**Review by eladiw — APPROVED** [review]
